src/utils/modexp/modexp_utils.cairo

```diff
@@ -0,0 +1,452 @@
+from starkware.cairo.common.cairo_builtins import BitwiseBuiltin
+from starkware.cairo.common.uint256 import (
+    Uint256,
+    uint256_add,
+    uint256_and,
+    uint256_check,
+    uint256_mul_div_mod,
+    uint256_mul,
+    uint256_eq,
+    uint256_lt,
+    uint256_shr,
+    uint256_sub,
+    uint256_unsigned_div_rem,
+)
+from starkware.cairo.common.bitwise import bitwise_and
+from starkware.cairo.common.registers import get_label_location
+
+// @title ModExpHelper Functions
+// @notice This file contains a selection of helper functions for modular exponentiation and gas cost calculation.
+// @author @dragan2234
+// @custom:namespace ModExpHelpers
+namespace ModExpHelpers {
+    const GAS_COST_MOD_EXP = 200;
+
+    func calculate_modexp_gas{range_check_ptr: felt, bitwise_ptr: BitwiseBuiltin*}(
+        b_size: Uint256, e_size: Uint256, m_size: Uint256, b: Uint256, e: Uint256, m: Uint256
```

Comment:
@ClementWalter you are right about the `b`,`e` and `m`. I just checked nethermind's implementation of this precompile and they are representing `b` `e` and `m` as `bytes[]` and using `gmp_lib` for arthimetics. This would maybe require different algorithm for modular exponentiation.

Reply:
@ClementWalter Thanks for pointing this out, now i realized the current implementation of ModExp in this PR only supports uint256, but actually bigint or `bytes` or `felt*` is needed. So this implementation is not correct. And there is no bigint library or implementation in cairo. There is this repo: https://github.com/NethermindEth/research-basic-Cairo-operations-big-integers
and this file: https://github.com/starkware-libs/cairo-lang/blob/master/src/starkware/cairo/common/cairo_secp/bigint.cairo

which are both not bigints with arbitrary size. What's interesting is that the modexp with big integers is needed for example in RSA verification as stated on EIP-198: https://github.com/ethereum/EIPs/blob/master/EIPS/eip-198.md

There is bigint library in rust for example: https://docs.rs/num-bigint/0.2.3/num_bigint/struct.BigUint.html


At least gas cost calcuation is done correctly as it operates with uint256 numbers, only for calcuation of gas cost. But the **bigint** modular exponentiation requires more work, as we would need to think how to abstract bigint of arbitrary size (for example 4096-bit number) with felts. I guess felt* is the way to go but then ModExp algorithm needs to be applied over that type